### DISCUSSION TITLE
Can't get Onsager analytical function to work for limiting case 2

### DISCUSSION THREAD

**Discussion by benpymer** [Part-II-Computational-Physics/.github#65] [General]
Hi!

I wrote this code to find the analytical solution for the average magnetisation in the case of no coupling

"""
for i in range(np.size(betaJs)):
    if betaJs[i] > (np.log(1 + np.sqrt(2)))/2:
        M_analytical_positive[i] = (1 - np.sinh((2*betaJs[i])**(-4)))**(1/8)
        M_analytical_negative[i] = -(1 - np.sinh((2*betaJs[i])**(-4)))**(1/8)
"""

However I seem to get a problem with M_analytical just after the critical point. The error message thrown is "RuntimeWarning: invalid value encountered in double_scalars" and my produced graph looks like this

![Screenshot (34)](https://user-images.githubusercontent.com/125975405/224068863-54c0fa23-f1e6-4714-acb4-7bb026bb43d0.png)

I having done some number crunching with a calculator, I think my issue is that $1 - sinh((2  \beta  J)^-4)$ is negative in the values of $\beta J$ where the graph is not showing, and therefore taking the eighth root returns an imaginary number, but I don't think this should be happening as I expect this is only meant to be imaginary in the region below the critical temperature where the magnetisation is 0 (in fact I assume this is why the solution has two regimes above and below a certain critical point). The only thing I can think of is that my critical point is in the wrong place, but I've checked it a lot and I think it should be right, although it's possible that I've coded a formula incorrectly. Does anyone know what's going wrong?

**Comment by benpymer** [comment]
Apologies for my formatting fails on GitHub btw but I hope it's clear what I'm asking!

**Comment by DLHvdH** [comment]
Below the critical point the magnetisation is expected to be zero, so you actually don't use the expression and just plot an array of zeros. The function is indeed a bit tricky in the sense that if you give it a value exactly at the betaJcritical (or just a tiny bit above) the function returns imaginary numbers. To get the function to work you thus may need to split 1) a zone below betaJcrit, 2) the positive branch, and 3) the negative branch. To get a full line, try to start plotting the non-zero branches at betaJcrit + 1e-10 or something.

**Reply by benpymer** [reply]
Hmmm ok. I did split into those branches already as you suggested, but I just included the code for the regime of under the critical temperature because I thought that's where the problem would be. My function seems to return imaginary numbers for about 0.1 of betaJ range above the betaJcrit value. Will plotting at betaJcrit + 1e-10 solve this? I'm not sure it will because I think betaJcrit works out to be about 0.44 and my calculator still gives me imaginary values of M_analytical for betaJ = 0.5? Am I wrong in where I think betaJcrit is?

**Reply by benpymer** [reply]
I'm using these betaJ values currently. I don't think any of them is within 1e-10 of 0.4406867935... = betaJcrit

betaJs = np.linspace(0.0,1.0,100)

**Reply by DLHvdH** [reply]
Your power -4 seems to be within the hypoberbolic sine function but it should be outside. Did you check that?

**Reply by benpymer** [reply]
Do you mean it's meant to be (sinh(x))^-4 rather than sinh(x^-4)? I was a bit confused by the notation in the question to be honest so I wouldn't be surprised if that was the issue

**Reply by benpymer** [reply]
I think that's fixed it. Thanks!
